docs/gni_sources.md

```diff
@@ -1,3 +1,9 @@
-Use of sources.gni should be limited to sources that would otherwise have circular dependencies that cannot be easily resolved, in other cases files should be added a new or existing BUILD.gn targets. The chromium ios code is a good model for separating out dependencies and sometimes makes use of interface/implementation patterns. A good example in desktop/android code is //chrome/browser/profiles target [https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/profiles/BUILD.gn;l=8](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/profiles/BUILD.gn;l=8) which has a minimal set of dependencies and the bulk of the implementation that would likely trigger a lot of circular deps is in profile_impl.cc in the //chrome/browser target. This reduces the number of sources/targets that either depend on //chrome/browser or have to be added to //chrome/browser sources list. Other strategies used to break //chrome dependencies in components can also sometimes be applied to //brave/browser code to avoid circular dependencies [https://www.chromium.org/developers/design-documents/cookbook/#dependency-inversion](https://www.chromium.org/developers/design-documents/cookbook/#dependency-inversion). Do not use `check_includes = false`.
+### sources.gni
 
-This does not mean that you cannot use sources.gni to inject sources into existing targets, but please try to avoid it when practical.
+Use of sources.gni to include sources in upstream targets should should generally be avoided. Whenever possible try to break circular dependencies see [Recipes for Breaking Chrome Dependencies](https://www.chromium.org/developers/design-documents/cookbook/#recipes-for-breaking-chrome-dependencies) and [https://www.chromium.org/developers/design-documents/cookbook/#dependency-inversion](https://www.chromium.org/developers/design-documents/cookbook/#dependency-inversion) for examples. An interface/implementation pattern can also often be used where header files and possibly some cc files are included in the direct dependency and the code that would cause the circular dependency is included in a higher level target like `//brave/browser` to ensure that the implementation code is always linked into the final output. See [tabs:tabs_public](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/tabs/BUILD.gn;l=12) and [tabs:impl](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/tabs/BUILD.gn;l=304) and [//chrome/browser impl dependency](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/BUILD.gn;l=4354). The chromium ios code is also a good model for separating out dependencies and sometimes makes use of interface/implementation patterns.
+
```

Comment:
i think people are mostly familiar with chrome vs components vs content dependency structure, so most confusion comes from /brave/browser target. There is still a very diverse mix of build.gn / source.gni / circular deps mess, so more detailed explanation (why we do it this way and not the other, with few positive examples (do like this) and negative (this is a problematic brave/browser/subdir) from brave-core would be extremely helpful for commiters

also a separate note about how we organize tests would be great to see

Reply:
I'lll add some background as well explaining how we go to the state we're in now

Reply:
not sure what you mean about tests?

Reply:
if you're talking about where tests go in general I think that's out of the scope of this document, but maybe there should be some more general doc that this becomes part of and includes that as well